[Fix #50] Make EndWith/StartWith autocorrects nil-safe

dduugg opened the pull request:
Makes the autocorrects for `Performance/EndWith` and `Performance/StartWith` safe to use with `nil` receivers, by invoking the autocorrected methods (`end_with?` and `start_with?`) with safe navigation where supported (ruby versions >= 2.3).

I've also enabled autocorrect for these cops in the default config. The autocorrect remains unsafe (particularly when correcting `Symbol` receivers, which do not support `end_with?` and `start_with?`, but this is captured by the `SafeAutoCorrect` setting. I'm happy to revert if the reviewer disagrees.

Ruby versions < 2.3 are not addressed, for reasons described in the linked ticket (impending end of support for the releases in rubocop core).

I've split the tests by ruby version where it mattered. This triggered the `RSpec/RepeatedExample` cop, which i evaded by some slight tweaks to the receiver name in the examples (`s`, `str`, `string`). Again, happy to take a different approach if the reviewer would like.

Fixes https://github.com/rubocop-hq/rubocop-performance/issues/50

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop-performance/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop-performance/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

dduugg commented:
As rubocop has [dropped support](https://github.com/rubocop-hq/rubocop/pull/7026) for Ruby 2.2, I'm happy to do the same in a separate PR for this repo, which will then simplify the diff here.

koic commented:
Sorry for the late decision and feedback. Support for Ruby 2.2 and lower drops. So the code for Ruby versions are unnecessary.

Also, I have not yet considered whether autocorrect should be enabled by default. Please wait a moment.

dduugg commented:
@koic Great! I've removed all version-specific code and rebased on latest master.

In case the data point is helpful, I applied these modified cops with autocorrect to a large code base. Only 2 of 127 corrections resulted in failed tests (we have high, but not perfect, test coverage). One was due to a `NoMethodError` from invocation on a `Symbol`, the other was in code that was explicitly checking for a `nil` (rather than falsy) response.

Let me know what you decide, and I'll update the PR if necessary. Thanks for your consideration.

koic commented:
I've investigated the history of auto-correct not safe for `Performance/StartWith` cop and `Performance/EndWith` cop.
https://github.com/rubocop-hq/rubocop/pull/3398

Unfortunately, I think that it does not seem safe auto-correct even if safe navigation operator is used. 

```diff
% cat example.rb
obj.match?(/\Aab/)

% rubocop --require rubocop-performance -a
Inspecting 1 file
C

Offenses:

example.rb:1:1: C: [Corrected] Performance/StartWith: Use String#start_with? instead of a regex match anchored to the beginning of the string.
obj.match?(/\Aab/)
^^^^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected, 1 offense corrected

% git diff
diff --git a/example.rb b/example.rb
index 8d21e67..4f59843 100644
--- a/example.rb
+++ b/example.rb
@@ -1 +1 @@
-obj.match?(/\Aab/)
+obj.start_with?('ab')
```

dduugg commented:
@koic We are in agreement that the autocorrect remains unsafe. Thus, `SafeAutoCorrect` remains false. I have enabled `AutoCorrect` by default, to be consistent with `RegexpMatch`. The latter is similarly unsafe:

```
$ cat example.rb
if obj.match(/ab/)
  do_stuff
end

$ rubocop --require rubocop-performance -a --only Performance/RegexpMatch
Inspecting 1 file
C

Offenses:

example.rb:1:4: C: [Corrected] Performance/RegexpMatch: Use match? instead of match when MatchData is not used.
if obj.match(/ab/)
   ^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected, 1 offense corrected

$ git diff
diff --git a/example.rb b/example.rb
index 05b7839..aa5bcb4 100644
--- a/example.rb
+++ b/example.rb
@@ -1,3 +1,3 @@
-if obj.match(/ab/)
+if obj.match?(/ab/)
   do_stuff
 end
```

If there is a reason to maintain the discrepancy in the default `AutoCorrect` setting between `RegexpMatch` and `EndWith`/`StartWith`, could you help me understand the distinction?

bbatsov commented:
@koic @dduugg Might be a good idea to wrap this PR up.

dduugg commented:
@bbatsov agreed, and thanks for the poke.

I'll close this out at the end of the week if the above question re: autocorrect discrepancies remains unanswered.

dduugg commented:
Timing out and closing, but still eager to land this if the maintainer(s) are interested.
